### PR TITLE
[libc/cmds] Replace qsort, enhance ls speed on FAT

### PR DESCRIPTION
**Pull request by ghaerr**
This PR should fix all the speed problems found in #1201.

The `ls` command was doing a `stat` system call for every file in a directory, which was quite slow on FAT. This enhanced version does not, unless required (e.g. using -l, -t, -F or -R).

In addition, `ls` was also doing quite a few `realloc` calls, which decreased its speed every 8 files displayed. Memory is allocated in 64-file chunks now for better throughput.

The `-t` option was also broken; it is now fixed.

Memory corruption was found in the libc `qsort` routine, and it was replaced with a version from BSD.
Long story short, the previous ELKS qsort routine had a reversed compare function, which required fixing some other commands that use qsort back to standard.

### CODE REVIEW

**Comment by toncho11** [comment]
Hi @ghaerr ,

Thank you for all the fixes! I will be able to test later on real hardware.

I think doing a sort everytime you do a listing is an overkill for a slow machine like 4.77 IBM PC and sometimes not needed.
I see the that the "-U" option is designed to prevent sorting in `ls`. Maybe you can add it?

**Comment by toncho11** [comment]
And have you seen this comment in the code:

```
28-May-2004 claudio@conectiva.com (Claudio Matsuoka)
 *		- Fixed sort direction, keeps qsort and strcmp consistent
 *		- Removed alias to 'dir' (doesn't seem to belong here)
 */
```

You can update the history note also.

**Comment by ghaerr** [comment]
Hello @toncho11,

> I will be able to test later on real hardware.

Thanks, I'm interested to hear how fast or slow the changes are on your 4.77 Mhz hardware.

> I see the that the "-U" option is designed to prevent sorting in ls. Maybe you can add it?

Sure, good idea.

> And have you seen this comment in the code:
> You can update the history note also.

Yes, I saw that and already had to update ls internally to keep qsort and strcmp consistent (again), so the note doesn't really need to change, thanks for noticing.

Thank you!

**Comment by Mellvik** [comment]
Incredible improvement, @ghaerr.
The `ls` the previously took 14 seconds, is now 0.75 secs. The one that took 44 secs, is now 1 second. The speed is now actually on par with DOS.

A great piece of work, @ghaerr. 
Thank you.

-M

```
# time ls
1000b     1001b     1023b     1024b     1025b     1100b     1200b     
128b      1459b     1999b     1byte     2000b     2001b     511b      
512b      513b      599b      600b      601b      767b      768b      
769b      999b      arp       aslookup  aup       banner    basename  
bc        cal       cat       chgrp     chmem     chmod     chown     
cksum     clear     clock     cmp       console   cp        cut       
date      dd        decomp16  df        diff      dirname   du        
echo      ed        env       false     fdisk     fdtest    fgrep     
file      find      fsck      ftp       ftp.c     ftp.o     ftpd      
ftpd.c    ftpd.o    ftpget    ftpput    getty     grep      hd        
head      httpd     httpget   init      kill      kilo      ktcp      
ln        login     logname   lp        ls        makeboot  makefile  
man       meminfo   mesg      mined     miniterm  mkdir     mkfat     
mkfifo    mkfs      mknod     more      mount     mouse     mv        
net       netstat   nslookup  od        p         partype   passwd    
poweroff  pp        printenv  proto     ps        pwd       pwdauth   
ramdisk   reboot    remsync   rm        rmdir     sash      sed       
sercat    sh        shutdown  sleep     sort      stty      sum       
sync      synctree  sys       tail      tar       tee       telnet    
telnetd   test      tget      time      tinyirc   touch     tr        
true      tty       umount    uniq      unreal16  urlget    uudecode  
uuencode  v         vi        vim       wc        which     who       
whoami    x         x1        x2        x3        x4        x5        
xaa       xab       xac       xacc      xad       xae       xaf       
xag       xah       xai       xaj       xak       xal       xam       
xan       xao       xap       xaq       xar       xargs     xas       
xat       xau       xav       xaw       xax       xay       xaz       
xba       xbb       xbc       xbd       xbe       xbf       xbg       
xbh       xbi       xbj       xbk       xbl       xbm       xbn       
xbo       xbp       xbq       xbr       xbs       xbt       xbu       
xbv       xbw       xbx       xby       xbz       xca       xcb       
xcc       xcd       xce       xcf       xcg       xch       xci       
xcj       xck       xcl       xcm       xcn       xco       xcp       
xcq       xcr       xcs       xct       xcu       xcv       xcw       
xcx       xcy       xcz       xda       xdb       xdc       xdd       
xde       xdf       xdg       xdh       xdi       xx        xx.sh     
y         zz        zz1       

Real    0m0.750s
# time ls zz
1000b     1001b     1023b     1024b     1025b     1100b     1200b     
128b      1459b     1999b     1byte     2000b     2001b     511b      
512b      513b      599b      600b      601b      767b      768b      
769b      999b      arp       aslookup  aup       banner    basename  
bc        cal       cat       chgrp     chmem     chmod     chown     
cksum     clear     clock     cmp       console   cp        cut       
date      dd        decomp16  df        diff      dirname   du        
echo      ed        env       false     fdisk     fdtest    fgrep     
file      find      fsck      ftp       ftp.c     ftp.o     ftpd      
ftpd.c    ftpd.o    ftpget    ftpput    getty     grep      hd        
head      httpd     httpget   init      kill      kilo      ktcp      
ln        login     logname   lp        ls        makeboot  makefile  
man       meminfo   mesg      mined     miniterm  mkdir     mkfat     
mkfifo    mkfs      mknod     more      mount     mouse     mv        
net       netstat   nslookup  od        p         partype   passwd    
poweroff  pp        printenv  proto     ps        pwd       pwdauth   
ramdisk   reboot    remsync   rm        rmdir     sash      sed       
sercat    sh        shutdown  sleep     sort      stty      sum       
sync      synctree  sys       tail      tar       tee       telnet    
telnetd   test      tget      time      tinyirc   touch     tr        
true      tty       umount    uniq      unreal16  urlget    uudecode  
uuencode  v         vi        vim       wc        which     who       
whoami    x         x1        x2        x3        x4        x5        
xaa       xab       xac       xacc      xad       xae       xaf       
xag       xah       xai       xaj       xak       xal       xam       
xan       xao       xap       xaq       xar       xargs     xas       
xat       xau       xav       xaw       xax       xay       xaz       
xba       xbb       xbc       xbd       xbe       xbf       xbg       
xbh       xbi       xbj       xbk       xbl       xbm       xbn       
xbo       xbp       xbq       xbr       xbs       xbt       xbu       
xbv       xbw       xbx       xby       xbz       xca       xcb       
xcc       xcd       xce       xcf       xcg       xch       xci       
xcj       xck       xcl       xcm       xcn       xco       xcp       
xcq       xcr       xcs       xct       xcu       xcv       xcw       
xcx       xcy       xcz       xda       xdb       xdc       xdd       
xde       xdf       xdg       xdh       xdi       

Real    0m1.020s
# 
```

**Comment by ghaerr** [comment]
> The ls the previously took 14 seconds, is now 0.75 secs. The one that took 44 secs, is now 1 second. The speed is now actually on par with DOS.

Great news!

Can you try an "ls -F" and/or "ls -F zz" to see how slow they are? (They will be slow - a stat will be required for each file, but I'm interested to see whether you still see the increasing slow-down per line or not).

Thank you!

**Comment by Mellvik** [comment]
Good idea, @ghaerr - 

and very interesting numbers:

# ls -F /mnt/tmp	12s
# cd /mnt/tmp
# ls -F		14s	(just like before)
# ls -F zz	33s	(25% faster than before, zz has the same files as tmp)
# ls -F /mnt/tmp/zz 34s
# ls -F /mnt/tmp/zz/zz1 58s (again, zz1 is a copy of zz, same files)

So listing the current directory is unchanged, relative and full-path directories are faster and depth is a killer.

And yes, the increasing slow-down is VERY noticeable.

-M

**Comment by ghaerr** [comment]
> So listing the current directory is unchanged, relative and full-path directories are faster and depth is a killer.
> And yes, the increasing slow-down is VERY noticeable.

Thanks again for the quick testing! 

My conclusion is that the `ls` rewrite definitely helped a lot by eliminating a needless `stat` per file - but that the `realloc` change didn't really do much, and that there's still an unknown reason as to why an increasing slow-down. I've added that to my list to continue to look at, as priority allows. At least large FAT directories can be listed much faster now.

Thank you!
